vyper/codegen/core.py

```diff
@@ -1152,14 +1152,17 @@ def clamp_dyn_array(ir_node, hi=None):
 
     assert (hi is not None) == _dirty_read_risk(ir_node)
 
-    # if the subtype is dynamic, the check will be performed in the recursion
-    if hi is not None and not t.abi_type.subtyp.is_dynamic():
+    if hi is not None:
         assert t.count < 2**64  # sanity check
 
         # note: this add does not risk arithmetic overflow because
         # length is bounded by count * elemsize.
         item_end = add_ofst(ir_node, _abi_payload_size(ir_node))
```

Comment:
feels off that we use `_abi_payload_size` - eg. in the test being added we'll trigger this path
```python
    if isinstance(ir_node.typ, DArrayT):
        return ["add", OFFSET, ["mul", get_dyn_array_count(ir_node), SCALE]]
```